### PR TITLE
Implement StageInstance

### PR DESCRIPTION
**Pull request by NCPlayz**
## Summary

Tracking discord/discord-api-docs#2898

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/stage_instance.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        'id',
+        'guild_id',
+        'channel_id',
+        'topic',
+        '_cs_guild',
+        '_cs_channel',
+    )
+
+    def __init__(self, *, state: ConnectionState, data: StageInstancePayload) -> None:
+        self._state: ConnectionState = state
```
Comment (Zomatree) [inline discord/stage_instance.py:71]:
```suggestion
        self._state = state
```

File: discord/stage_instance.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    from .types.stage_instance import StageInstance as StageInstancePayload
+    from .state import ConnectionState
+    from .channel import StageChannel
+    from .guild import Guild
+
+
+class StageInstance(Hashable):
+    """Represents a stage instance of a stage channel in a guild.
+
+    .. versionadded:: 2.0
```
Comment (Rapptz) [inline discord/stage_instance.py:46]: This is missing the operations container.

File: discord/channel.py [hunk 3]
```diff
@@ -640,7 +641,7 @@ def voice_states(self):
         return {key: value for key, value in self.guild._voice_states.items() if value.channel.id == self.id}
 
     @utils.copy_doc(discord.abc.GuildChannel.permissions_for)
-    def permissions_for(self, member):
+    def permissions_for(self, member: Member, /) -> Permissions:
```
Comment (Rapptz) [inline discord/channel.py:644]:
```suggestion
    def permissions_for(self, member: Union[Role, Member], /) -> Permissions:
```

File: discord/channel.py [hunk 4]
```diff
@@ -875,20 +876,114 @@ def _update(self, guild, data):
...
+
+        Create a stage instance.
+
+        .. versionadded:: 2.0
+
+        Parameters
+        -----------
+        topic: :class:`str`
+            The stage instance's topic.
+        privacy_level: :class:`PrivacyLevel`
```
Comment (Zomatree) [inline discord/channel.py:936]:
```suggestion
        privacy_level: Optional[:class:`PrivacyLevel`]
```

Comment (NCPlayz) [reply discord/channel.py:936]: Swapped to using utils.MISSING since this can't actually be set as None.

File: discord/types/stage_instance.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from typing import Literal, TypedDict
+
+from .snowflake import Snowflake
+
+PrivacyLevel = Literal[1, 2]
```
Comment (Rapptz) [inline discord/types/stage_instance.py:29]:
```suggestion
StagePrivacyLevel = Literal[1, 2]
```

File: discord/types/stage_instance.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+"""
```
Comment (Rapptz) [inline discord/types/stage_instance.py:1]: Maybe move these typings to `discord.types.channel` instead its own dedicated file? I'm not sure these are special enough from channels to warrant this separation atm.

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/stage_instance.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    async def edit(self, *, topic: str = None, privacy_level: PrivacyLevel = MISSING) -> None:
+        """|coro|
+
+        Edits the stage instance.
+
+        Parameters
+        -----------
+        topic: :class:`str`
+            The stage instance's new topic.
```
[the remark this answers is not in the document]
Comment (NCPlayz) [reply discord/stage_instance.py:122]: It, unfortunately, does not. It seems that the `topic` attribute is crucial to a stage instance.

File: discord/channel.py [hunk 8]
```diff
@@ -875,20 +876,116 @@ def _update(self, guild, data):
...
```
Comment (Rapptz) [inline discord/channel.py:936]:
```suggestion
        privacy_level: :class:`StagePrivacyLevel`
```

File: discord/stage_instance.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    -----------
+    id: :class:`int`
+        The stage instance's ID.
+    guild: :class:`Guild`
+        The guild that the stage instance is running in.
+    channel_id: :class:`int`
+        The ID of the channel that the stage instance is running in.
+    topic: :class:`str`
+        The topic of the stage instance.
+    privacy_level: :class:`PrivacyLevel`
```
Comment (Rapptz) [inline discord/stage_instance.py:74]:
```suggestion
    privacy_level: :class:`StagePrivacyLevel`
```

File: discord/stage_instance.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+    async def edit(self, *, topic: str = MISSING, privacy_level: StagePrivacyLevel = MISSING) -> None:
+        """|coro|
+
+        Edits the stage instance.
+
+        Parameters
+        -----------
+        topic: :class:`str`
+            The stage instance's new topic.
+        privacy_level: :class:`PrivacyLevel`
```
Comment (Rapptz) [inline discord/stage_instance.py:123]:
```suggestion
        privacy_level: :class:`StagePrivacyLevel`
```

File: docs/api.rst [hunk 11]
```diff
@@ -2029,6 +2055,24 @@ of :class:`enum.Enum`.
 
         Represents full camera video quality.
 
+.. class:: PrivacyLevel
```
Comment (Rapptz) [inline docs/api.rst:2058]:
```suggestion
.. class:: StagePrivacyLevel
```

File: docs/api.rst [hunk 12]
```diff
@@ -827,6 +827,32 @@ to handle it, which defaults to print a traceback and ignoring the exception.
...
+    .. versionadded:: 2.0
+
+    :param stage_instance: The stage instance that was created or deleted.
+    :type stage_instance: :class:`StageInstance`
+
+.. function:: on_stage_instance_update(before, after)
+
+    Called when a :class:`StageInstance` is updated.
+
+    .. versionadded:: 2.0
```
Comment (Rapptz) [inline docs/api.rst:844]: Needs to be below the paragraphs but above the parameter list.

**Review by Rapptz — CHANGES_REQUESTED** [review]
A few minor nits.

**Review by Rapptz — APPROVED** [review]

**Pull request merged**